### PR TITLE
Refactor nse

### PR DESCRIPTION
**Pull request by strengejacke**

### CODE REVIEW

**Comment by strengejacke** [comment]
``` r
library(datawizard)
data_addsuffix(iris, "HA", starts_with("Sepal")) |> head()
#>   Sepal.LengthHA Sepal.WidthHA Petal.Length Petal.Width Species
#> 1            5.1           3.5          1.4         0.2  setosa
#> 2            4.9           3.0          1.4         0.2  setosa
#> 3            4.7           3.2          1.3         0.2  setosa
#> 4            4.6           3.1          1.5         0.2  setosa
#> 5            5.0           3.6          1.4         0.2  setosa
#> 6            5.4           3.9          1.7         0.4  setosa
data_addsuffix_test(iris, "HA", starts_with("Sepal")) |> head()
#> Warning in eval(p): restarting interrupted promise evaluation

#> Warning in eval(p): restarting interrupted promise evaluation
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1          5.1         3.5          1.4         0.2  setosa
#> 2          4.9         3.0          1.4         0.2  setosa
#> 3          4.7         3.2          1.3         0.2  setosa
#> 4          4.6         3.1          1.5         0.2  setosa
#> 5          5.0         3.6          1.4         0.2  setosa
#> 6          5.4         3.9          1.7         0.4  setosa
```

<sup>Created on 2022-03-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
(related code: https://github.com/easystats/datawizard/pull/115/commits/eeb5ab4eea33b61e67cded8347d292e7b1b56de1)

File: R/test_nse.R [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    fixed <- evaluated_pattern$fixed
+  }
+
+  # seems to be no valid column name or index, so try to grep
+  if (isFALSE(fixed)) {
+    select <- colnames(data)[grepl(select, colnames(data), ignore.case = ignore_case)]
+  }
+
+  # load again
+  .attach_packages(conflicting_packages)
```
Comment (mattansb) [inline R/test_nse.R:41]: This is a problem is e.g., `poorman` was loaded before `dplyr`, so for the used before running this function dplyr masks poorman, but after running it, poorman masks dplyr...

Comment (strengejacke) [reply R/test_nse.R:41]: That's unrelated. The problem is that NSE is not working when passing down `select` to a helper function:

``` r
library(datawizard)
data_addsuffix(iris, "HA", col_starts_with("Sepal")) |> head()
#>   Sepal.LengthHA Sepal.WidthHA Petal.Length Petal.Width Species
#> 1            5.1           3.5          1.4         0.2  setosa
#> 2            4.9           3.0          1.4         0.2  setosa
#> 3            4.7           3.2          1.3         0.2  setosa
#> 4            4.6           3.1          1.5         0.2  setosa
#> 5            5.0           3.6          1.4         0.2  setosa
#> 6            5.4           3.9          1.7         0.4  setosa
data_addsuffix_test(iris, "HA", col_starts_with("Sepal")) |> head()
#> Warning in eval(p): restarting interrupted promise evaluation

#> Warning in eval(p): restarting interrupted promise evaluation
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1          5.1         3.5          1.4         0.2  setosa
#> 2          4.9         3.0          1.4         0.2  setosa
#> 3          4.7         3.2          1.3         0.2  setosa
#> 4          4.6         3.1          1.5         0.2  setosa
#> 5          5.0         3.6          1.4         0.2  setosa
#> 6          5.4         3.9          1.7         0.4  setosa
```

<sup>Created on 2022-03-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
It's really the NSE that is failing here:

``` r
library(datawizard)

# data_addsuffix_test doesn't work

data_addsuffix(iris, "HA", starts_with("Sepal")) |> colnames()
#> [1] "Sepal.LengthHA" "Sepal.WidthHA"  "Petal.Length"   "Petal.Width"   
#> [5] "Species"
data_addsuffix_test(iris, "HA", starts_with("Sepal")) |> colnames()
#> Warning in eval(p): restarting interrupted promise evaluation

#> Warning in eval(p): restarting interrupted promise evaluation
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

data_addsuffix(iris, "HA", Species) |> colnames()
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "SpeciesHA"
data_addsuffix_test(iris, "HA", Species) |> colnames()
#> Warning in eval(p): restarting interrupted promise evaluation

#> Warning in eval(p): restarting interrupted promise evaluation
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"


# data_addsuffix_test works

test <- c("Species", "Petal.Width")
data_addsuffix(iris, "HA", test) |> colnames()
#> [1] "Sepal.Length"  "Sepal.Width"   "Petal.Length"  "Petal.WidthHA"
#> [5] "SpeciesHA"
data_addsuffix_test(iris, "HA", test) |> colnames()
#> [1] "Sepal.Length"  "Sepal.Width"   "Petal.Length"  "Petal.WidthHA"
#> [5] "SpeciesHA"

data_addsuffix(iris, "HA", "Species") |> colnames()
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "SpeciesHA"
data_addsuffix_test(iris, "HA", "Species") |> colnames()
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "SpeciesHA"

data_addsuffix(iris, "HA", ~Species + Petal.Width) |> colnames()
#> [1] "Sepal.Length"  "Sepal.Width"   "Petal.Length"  "Petal.WidthHA"
#> [5] "SpeciesHA"
data_addsuffix_test(iris, "HA", ~Species + Petal.Width) |> colnames()
#> [1] "Sepal.Length"  "Sepal.Width"   "Petal.Length"  "Petal.WidthHA"
#> [5] "SpeciesHA"

data_addsuffix(iris, "HA", c("Species", "Petal.Width")) |> colnames()
#> [1] "Sepal.Length"  "Sepal.Width"   "Petal.Length"  "Petal.WidthHA"
#> [5] "SpeciesHA"
data_addsuffix_test(iris, "HA", c("Species", "Petal.Width")) |> colnames()
#> [1] "Sepal.Length"  "Sepal.Width"   "Petal.Length"  "Petal.WidthHA"
#> [5] "SpeciesHA"

testfun1 <- function(wantit) {
  data_addsuffix(iris, "HA", wantit) |> colnames()
}
testfun2 <- function(wantit) {
  data_addsuffix_test(iris, "HA", wantit) |> colnames()
}
testfun1(c("Species", "Petal.Width"))
#> [1] "Sepal.Length"  "Sepal.Width"   "Petal.Length"  "Petal.WidthHA"
#> [5] "SpeciesHA"
testfun2(c("Species", "Petal.Width"))
#> [1] "Sepal.Length"  "Sepal.Width"   "Petal.Length"  "Petal.WidthHA"
#> [5] "SpeciesHA"
```

My question is: how can we refactor the code for evaluating the `select` argument and put it in an own function, to avoid duplicated code?

**Comment by strengejacke** [comment]
@bwiernik maybe you have an idea here?

**Comment by mattansb** [comment]
Right, it's failing because you are evaluating the argument - so it isn't a promise anymore.

I would change the order of perations:

1. deparse `select`
2. Does it match any of the special NSE operations we support?
3. If not, is it a character? (starts and ends with " or ')
4. If neither 2 or 3, evaluate it as it is some argument passed down from else where.

**Comment by strengejacke** [comment]
Not sure how this would work... Feel free to modify `test_fun()` to make the above examples work for `data_addsuffix_test()`.

**Comment by mattansb** [comment]
Ta da!

``` r
library(datawizard)
library(testthat)
#> Warning: package 'testthat' was built under R version 4.1.2

expect_equal(
  data_addsuffix(iris, "HA", starts_with("Sepal")) |> colnames(),  
  data_addsuffix_test(iris, "HA", starts_with("Sepal")) |> colnames()
)

expect_equal(
  data_addsuffix(iris, "HA", Species) |> colnames(),
  data_addsuffix_test(iris, "HA", Species) |> colnames()
)



test <- c("Species", "Petal.Width")
expect_equal(
  data_addsuffix(iris, "HA", test) |> colnames(),
  data_addsuffix_test(iris, "HA", test) |> colnames()
)

expect_equal(
  data_addsuffix(iris, "HA", "Species") |> colnames(),
  data_addsuffix_test(iris, "HA", "Species") |> colnames()  
)


expect_equal(
  data_addsuffix(iris, "HA", ~Species + Petal.Width) |> colnames(),
  data_addsuffix_test(iris, "HA", ~Species + Petal.Width) |> colnames()  
)

expect_equal(
  data_addsuffix(iris, "HA", c("Species", "Petal.Width")) |> colnames(),
  data_addsuffix_test(iris, "HA", c("Species", "Petal.Width")) |> colnames()
)


testfun1 <- function(wantit) {
  data_addsuffix(iris, "HA", wantit) |> colnames()
}
testfun2 <- function(wantit) {
  data_addsuffix_test(iris, "HA", wantit) |> colnames()
}

expect_equal(
  testfun1(c("Species", "Petal.Width")),
  testfun2(c("Species", "Petal.Width"))
)
```

<sup>Created on 2022-03-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

File: R/test_nse.R [hunk 2]
```diff
@@ -17,7 +17,7 @@ test_fun <- function(select, data, exclude, ignore_case) {
   # in case pattern is a variable from another function call...
   p <- try(eval(select), silent = TRUE)
   if (inherits(p, c("try-error", "simpleError"))) {
-    p <- substitute(select)
+    p <- substitute(select, env = parent.frame())
```
Comment (strengejacke) [inline R/test_nse.R:20]: what

Comment (mattansb) [reply R/test_nse.R:20]: Gotta tell R where to look!

**Comment by strengejacke** [comment]
Cool, thanks! Moving to #119

**Comment by bwiernik** [comment]
Sorry, just saw this.

**Pull request closed without merging**